### PR TITLE
Add coherence and Diff logging for LDA

### PR DESCRIPTION
**Pull request by parulsethi**
Added a method to log coherence during training.

### CODE REVIEW

**Comment by tmylk** [comment]
Please update coherence tutorial notebook with this new feature.
Also add a separate notebook that graphs both coherence and perplexity through training simultaneously. See examples in #1243 early versions

**Comment by menshikh-iv** [comment]
I'm worried that there are a lot of arguments in the LDA constructor and they become even bigger. We need to think about this problem.

File: gensim/models/ldamodel.py [hunk 1]
```diff
@@ -527,6 +528,15 @@ def log_perplexity(self, chunk, total_docs=None):
                     (perwordbound, np.exp2(-perwordbound), len(chunk), corpus_words))
         return perwordbound
 
+    def log_coherence(self, model, chunk):
+        """
+        Log 'u_mass' coherence using the `chunk` of documents as evaluation corpus.
+        """
+        cm = gensim.models.CoherenceModel(model=model, corpus=chunk, dictionary=self.id2word, coherence='u_mass')
+        corpus_words = sum(cnt for document in chunk for _, cnt in document)
+        logger.info("%.3f coherence estimate based on a held-out corpus of %i documents with %i words", cm.get_coherence(), len(chunk), corpus_words)
```
Comment (menshikh-iv) [inline gensim/models/ldamodel.py]: Calculate coherence only once and save it to variable

File: gensim/models/ldamodel.py [hunk 2]
```diff
@@ -527,6 +528,15 @@ def log_perplexity(self, chunk, total_docs=None):
                     (perwordbound, np.exp2(-perwordbound), len(chunk), corpus_words))
         return perwordbound
 
+    def log_coherence(self, model, chunk):
```
Comment (menshikh-iv) [inline gensim/models/ldamodel.py]: No needed to pass model at this method, you already have a model as `self`

**Comment by parulsethi** [comment]
Updated to add diff logging.

File: gensim/models/ldamodel.py (hunk 1 above)
Comment (parulsethi) [inline gensim/models/ldamodel.py]: Done

File: gensim/models/ldamodel.py (hunk 2 above)
Comment (parulsethi) [inline gensim/models/ldamodel.py]: Done

File: gensim/models/ldamodel.py [hunk 3]
```diff
@@ -195,7 +197,9 @@ def __init__(self, corpus=None, num_topics=100, id2word=None,
                  alpha='symmetric', eta=None, decay=0.5, offset=1.0,
                  eval_every=10, iterations=50, gamma_threshold=0.001,
                  minimum_probability=0.01, random_state=None, ns_conf={},
-                 minimum_phi_value=0.01, per_word_topics=False):
+                 minimum_phi_value=0.01, per_word_topics=False, coherence='u_mass',
+                 coherence_texts=None, coherence_window_size=None, coherence_topn=10,
```
Comment (menshikh-iv) [inline gensim/models/ldamodel.py]: use `metrics` from another PR

File: gensim/models/ldamodel.py [hunk 4]
```diff
@@ -529,9 +569,33 @@ def log_perplexity(self, chunk, total_docs=None):
...
+        prev_epoch = epoch - 1
+        if epoch == 0:
+            prev_epoch = "initial random model"
+        logger.info("Topic difference between %s and %s epoch: %s", prev_epoch, epoch, diff_diagonal)
+        return diff_diagonal
+
     def update(self, corpus, chunksize=None, decay=None, offset=None,
                passes=None, update_every=None, eval_every=None, iterations=None,
-               gamma_threshold=None, chunks_as_numpy=False):
+               gamma_threshold=None, chunks_as_numpy=False, coherence=None,
```
Comment (menshikh-iv) [inline gensim/models/ldamodel.py]: use `metrics` from another PR

**Comment by menshikh-iv** [comment]
Continued in #1399

**Pull request closed without merging**